services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/DefaultS3TransferManager.java

```diff
@@ -460,10 +461,19 @@ private static void handleException(CompletableFuture<CompletedFileDownload> ret
                                         CompletableFuture<TransferProgress> progressFuture,
                                         CompletableFuture<DownloadFileRequest> newDownloadFileRequestFuture,
                                         Throwable throwable) {
-        SdkClientException exception = SdkClientException.create("Failed to resume the request", throwable);
-        returnFuture.completeExceptionally(exception);
-        progressFuture.completeExceptionally(exception);
-        newDownloadFileRequestFuture.completeExceptionally(exception);
+        Throwable cause = throwable instanceof CompletionException ? throwable.getCause() : throwable;
+
+        if (cause instanceof SdkException || cause instanceof Error) {
+            returnFuture.completeExceptionally(cause);
+            progressFuture.completeExceptionally(cause);
+            newDownloadFileRequestFuture.completeExceptionally(cause);
+        } else {
```

Comment:
Also can we reduce the code to below one , What do you say  ?

```java

    private static void handleException2(CompletableFuture<CompletedFileDownload> returnFuture,
                                        CompletableFuture<TransferProgress> progressFuture,
                                        CompletableFuture<DownloadFileRequest> newDownloadFileRequestFuture,
                                        Throwable throwable) {
        
        Throwable exceptionCause = throwable instanceof CompletionException ? throwable.getCause() : throwable;
        
        Throwable propagatedException = exceptionCause instanceof SdkException || exceptionCause instanceof Error 
                                        ? exceptionCause
                                        : SdkClientException.create("Failed to resume the request", exceptionCause);

        returnFuture.completeExceptionally(propagatedException);
        progressFuture.completeExceptionally(propagatedException);
        newDownloadFileRequestFuture.completeExceptionally(propagatedException);

    }
```

Reply:
Agreed, looks much better